### PR TITLE
[Enhance] Add runner test case.

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

All unit tests need to build runner could inherit `TestRunnerTestCase`.

## Modification

1. A lot of unit test needs to build a `Runner` instance, `RunnerTestCase` provides a simple way to build runner with iter based or epoch-based config.
2. Building a `Runner` will leave a lot of global variable, `RunnerTestCase` will clean then after each unit test.
3. `RunnerTestCase` will deinitialize the distributed enviroment.
4. `Registry` instance like `MODELS` and `HOOKS` are also global variables,  `RunnerTestCase` will also clean them after each unit test,

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
5. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
6. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
7. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/testing/runner_test_case.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+    def process(self, data_batch, predictions):
+        result = {'acc': 1}
+        self.results.append(result)
+
+    def compute_metrics(self, results):
+        return dict(acc=1)
+
+
+class RunnerTestCase(TestCase):
```
Comment (C1rN09) [inline mmengine/testing/runner_test_case.py:82]: I think it's better to add some docstring here, which may help understand what it's designed for and how to use it.

File: mmengine/testing/runner_test_case.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        self.epoch_based_cfg = Config(epoch_based_cfg)
+        self.iter_based_cfg: Config = copy.deepcopy(self.epoch_based_cfg)
+        self.iter_based_cfg.train_dataloader = dict(
+            dataset=dict(type='ToyDataset'),
+            sampler=dict(type='InfiniteSampler', shuffle=True),
+            batch_size=3,
+            num_workers=0)
+        self.iter_based_cfg.log_processor = dict(by_epoch=False)
+
+        # prepare iter based cfg.
```
Comment (zhouzaida) [inline mmengine/testing/runner_test_case.py:134]:
```suggestion

        # prepare iter based cfg
        self.iter_based_cfg: Config = copy.deepcopy(self.epoch_based_cfg)
        self.iter_based_cfg.train_dataloader = dict(
            dataset=dict(type='ToyDataset'),
            sampler=dict(type='InfiniteSampler', shuffle=True),
            batch_size=3,
            num_workers=0)
        self.iter_based_cfg.log_processor = dict(by_epoch=False)
```

File: tests/test_runner/test_runner.py [hunk 3]
```diff
@@ -1487,7 +1511,7 @@ def before_val_iter(self, runner, batch_idx, data_batch=None):
         val_interval_results = []
         val_interval_targets = [5] * 10 + [2] * 2
 
-        @HOOKS.register_module()
+        @HOOKS.register_module(force=True)
```
Comment (zhouzaida) [inline tests/test_runner/test_runner.py:1514]: Why do we need to set `force=True`?

File: tests/test_model/test_model_utils.py [hunk 4]
```diff
@@ -79,7 +79,7 @@ def __init__(self, model):
 
         pass
 
-    CHILD_REGISTRY.register_module(module=CustomModelWrapper)
+    CHILD_REGISTRY.register_module(module=CustomModelWrapper, force=True)
```
Comment (zhouzaida) [inline tests/test_model/test_model_utils.py:82]: Same as above

File: mmengine/testing/runner_test_case.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+    def build_runner(self, cfg):
+        cfg.experiment_name = self.experiment_name
+        runner = Runner.from_cfg(cfg)
+        return runner
+
+    @property
+    def experiment_name(self):
+        return f'{self._testMethodName}_{time.time()} + ' \
+               f'{uuid4()}'
```
Comment (zhouzaida) [inline mmengine/testing/runner_test_case.py:164]: Add a comment to describe why we need to append `uuid4()`.

File: tests/test_runner/test_runner.py (hunk 3 above)
Comment (HAOCHENYE) [reply tests/test_runner/test_runner.py:1514]: Some toy classes with the same name are defined in different unit tests. If `force=False`, the class defined in current unit test will not be registered. If we refactor the unit test and make sure each unit test will clean the registered class, there is no need to set `force=True` anymore.

File: mmengine/testing/runner_test_case.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    def compute_metrics(self, results):
+        return dict(acc=1)
+
+
+class RunnerTestCase(TestCase):
+    """A test case to build runner easily.
+
+    `RunnerTestCase` will do the following things:
+
+    1. Registers a toy model, toy metric, toy dataset, which can be used to
```
Comment (zhouzaida) [inline mmengine/testing/runner_test_case.py:87]:
```suggestion
    1. Registers a toy model, a toy metric, and a toy dataset, which can be used to
```

**Review by zhouzaida — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**